### PR TITLE
refactor(snapbox): Changes to normalization

### PR DESCRIPTION
**Pull request by Muscraft**
This is a part of #92.

This PR is to help move forward the effort with supporting structured data formats. Part of this effort showed that there needed to be a better way to specify what normalization operation was wanted and later to unify what the operations are doing. 

- Abstract lower level `map_text` calls as `normalize` with operation types
  - By providing build blocks before, we were hoping to allow people to plug in their own normalization routines.  However, when we add structured data support, it will need to operate on internal data initially as we work out details, so we're making all of the normalization logic private, just allowing the caller to which built-in normalization can be done.
- change `as_str` to `try_string` and update its return to be `Option<String>`
  - The original intent of `as_str` was for outputting Diffs. This had to be changed to `String` since there are some formats that have borrowing issues with `&str`. By changing to `String` it allows support for more formats

### CODE REVIEW

File: crates/snapbox/src/data.rs [hunk 1]
```diff
@@ -110,10 +107,10 @@ impl Data {
     /// Return the underlying `str`
     ///
     /// Note: this will not inspect binary data for being a valid `str`.
-    pub fn as_str(&self) -> Option<&str> {
+    pub fn try_string(&self) -> Option<String> {
```
Comment (epage) [inline crates/snapbox/src/data.rs:110]: > change as_str to try_string and update its return to be Option<String>
> - This was to allow for different data formats that don't have an easy way to be converted to &str

This is in the checklist as

> as_str should be renamed to try_string and it should output Option<String>. This is because some data formats cannot be turned into a &str without borrowing issues

What neither captures is what is the intention behind `as_str` / `try_string`.  What role is it filling?  I can go through and refresh myself on that but the PR should do that.

Comment (Muscraft) [reply crates/snapbox/src/data.rs:110]: I updated the PR description to provide what `as_str` is for and why changing it is needed

Comment (epage) [inline crates/snapbox/src/data.rs:110]: `try_string` doesn't make the intent clear for structured data.  What about `render` or something?

File: crates/snapbox/src/assert.rs [hunk 2]
```diff
@@ -197,17 +195,20 @@ impl Assert {
...
+                .normalize(NormalizeMatches::new(&self.substitutions, ok));
+        } else {
             actual = actual
-                .try_coerce(format)
-                .map_text(|s| self.normalize_text(s))
-                .map_text(|t| self.substitutions.normalize(t, expected));
+                .try_coerce(ok.format())
+                .normalize(NormalizeNewlines)
+                .normalize(NormalizeMatches::new(&self.substitutions, ok));
         }
```
Comment (epage) [inline crates/snapbox/src/assert.rs:212]: This doesn't do a good job communicating what is happening.  What if instead we just dropped `NormalizeText` and had all callers of it call the parts its made up of?

File: src/schema.rs [hunk 3]
```diff
@@ -64,7 +64,7 @@ impl TryCmd {
                 sequence
             } else if ext == std::ffi::OsStr::new("trycmd") || ext == std::ffi::OsStr::new("md") {
                 let raw = crate::Data::read_from(path, Some(snapbox::DataFormat::Text))?
-                    .map_text(snapbox::utils::normalize_lines)
+                    .normalize(snapbox::NormalizeNewlines)
```
Comment (epage) [inline src/schema.rs:67]: This is an abuse of `Data` that we were supposed to move off of before this change

Comment (epage) [reply src/schema.rs:67]: In general, I'd recommend doing a scrub of changes to this file to make sure we are only using Data when we should and only using `try_string` when we should

File: crates/snapbox/src/data.rs (hunk 1 above)
Comment (Muscraft) [reply crates/snapbox/src/data.rs]: I switched to `render` since it makes the most sense and describes what it is for

File: src/schema.rs [hunk 4]
```diff
@@ -1202,9 +1199,9 @@ $ rust-cmd1
         let input = "One\nTwo\nThree";
         let line_nums = 2..3;
         let replacement = "World\n";
-        let expected = Data::text("One\nWorld\nThree");
+        let expected = "One\nWorld\nThree";
 
-        let mut actual = Data::text(input);
+        let mut actual = input.to_string();
```
Comment (epage) [inline src/schema.rs:1204]: `to_string`
- Obscures intent
- Is sloppier, making it harder to safely refactor code since it will work with any `display`

In this case, I would use `.to_owned`

File: crates/snapbox/src/data.rs [hunk 5]
```diff
@@ -110,10 +107,10 @@ impl Data {
     /// Return the underlying `str`
     ///
     /// Note: this will not inspect binary data for being a valid `str`.
-    pub fn as_str(&self) -> Option<&str> {
+    pub fn render(&self) -> Option<String> {
```
Comment (epage) [inline crates/snapbox/src/data.rs:110]: This is unrelated to the rest of the commit.  Could you split this into its own commit?

File: crates/snapbox/src/lib.rs [hunk 6]
```diff
@@ -114,6 +114,7 @@ pub use action::DEFAULT_ACTION_ENV;
 pub use assert::Assert;
 pub use data::Data;
 pub use data::DataFormat;
+pub use data::{Normalize, NormalizeMatches, NormalizeNewlines, NormalizePaths};
```
Comment (epage) [inline crates/snapbox/src/lib.rs:117]: side note: Hmm, one advantage of an `enum` is its a bit less of a mouthful everywhere.

File: crates/snapbox/src/data.rs [hunk 7]
```diff
@@ -206,6 +203,64 @@ impl<'s> From<&'s str> for Data {
...
+                let lines = crate::utils::normalize_paths(&text);
+                Data::text(lines)
+            }
+        }
+    }
+}
+
+pub struct NormalizeMatches<'a> {
+    substitutions: &'a crate::Substitutions,
+    pattern: Data,
```
Comment (epage) [inline crates/snapbox/src/data.rs:238]: If you are already borrowing, how come you don't borrow `pattern`?

File: src/schema.rs [hunk 8]
```diff
@@ -143,10 +146,7 @@ impl TryCmd {
                         .expected_stdout_source
                         .clone()
                         .expect("always present for .trycmd");
-                    let mut stdout = stdout
-                        .as_str()
-                        .expect("already converted to Text")
-                        .to_owned();
+                    let mut stdout = stdout.render().expect("already converted to Text");
```
Comment (epage) [inline src/schema.rs:149]: Should probably be updated to "at least Text"

File: crates/snapbox/src/path.rs [hunk 9]
```diff
@@ -257,15 +260,14 @@ impl PathDiff {
                         crate::Data::read_from(&actual_path, None).map_err(Self::Failure)?;
 
                     let expected = crate::Data::read_from(&expected_path, None)
-                        .map(|d| d.map_text(crate::utils::normalize_lines))
+                        .map(|d| d.normalize(NormalizeNewlines))
                         .map_err(Self::Failure)?;
 
-                    if let (Some(expected), format) = (expected.as_str(), expected.format()) {
```
Comment (epage) [inline crates/snapbox/src/path.rs:263]: side note: The dropping of the `if`s would have ideally been a separate commit so they don't get lost as we need to consider how safe this is.

I *think* it should be safe

Comment (Muscraft) [reply crates/snapbox/src/path.rs:263]: I moved the if statements into their own commit, so we can look at them individually

File: crates/snapbox/src/lib.rs (hunk 6 above)
Comment (Muscraft) [reply crates/snapbox/src/lib.rs:117]: I didn't change to an enum but if it seems like a better fit I can look certainly try to implement it

Comment (epage) [reply crates/snapbox/src/lib.rs:117]: Can't remember what trade offs we considered.  I marked this as a side note because I wanted to be clear this wasn't me prescriptively saying we should switch.  I leave it to you who has been working on this and probably thought through the different angles.

File: crates/snapbox/src/assert.rs [hunk 10]
```diff
@@ -201,13 +201,13 @@ impl Assert {
         // On `expected` being an error, make a best guess
         if let (Some(expected), format) = expected
             .as_ref()
-            .map(|d| (d.as_str(), d.format()))
-            .unwrap_or((Some(""), DataFormat::Text))
+            .map(|d| (d.render(), d.format()))
```
Comment (epage) [inline crates/snapbox/src/assert.rs]: This doesn't look correct.  I bet if you move the `if` removal first, this won't matter anymore

File: crates/snapbox/src/assert.rs [hunk 11]
```diff
@@ -201,13 +201,13 @@ impl Assert {
...
+            .unwrap_or((Some("".to_string()), DataFormat::Text))
```
Comment (epage) [inline crates/snapbox/src/assert.rs]: Again, `to_string` makes refactoring harder

File: crates/snapbox/src/assert.rs [hunk 12]
```diff
@@ -197,22 +197,15 @@ impl Assert {
...
-                actual = actual.normalize(NormalizePaths);
-            }
-            actual = actual
-                .normalize(NormalizeNewlines)
-                .normalize(NormalizeMatches::new(
-                    &self.substitutions,
-                    &crate::Data::text(expected),
-                ));
+
+        let ok = expected.clone().unwrap_or_default();
```
Comment (epage) [inline crates/snapbox/src/assert.rs:201]: Why is it cloned?

Why are we defaulting the whole thing?

Comment (Muscraft) [reply crates/snapbox/src/assert.rs:201]: `clone` is because we can't use `as_ref` since it would make the `unwarp` portion need to be ref as well. As for `unwrap_or_default` I believe it would make it `Data::new()` which is `Text` but I can see now that an explicit `unwrap_or(Data::new())` might be the better idea.

Comment (epage) [reply crates/snapbox/src/assert.rs:201]: > clone is because we can't use as_ref since it would make the unwarp portion need to be ref as well.

You could create a dummy variable and unwrap to that.

but really, this is contingent on the unwrap

> As for unwrap_or_default I believe it would make it Data::new() which is Text but I can see now that an explicit unwrap_or(Data::new()) might be the better idea.

If instead we do `expected.as_ref().map(|e| e.format()).unwrap_or_default())` and conditionalize the `NormalizeMatches` on there being an `expected`, then we don't need the clone.  The performance of it probably doesn't matter but this would help clarify intent.
